core/src/query.rs

```diff
@@ -1,77 +1,551 @@
+                            // the top-level any_of and set the top-level items to
+                            // None.
+                            BsonTypeName::Array => {
+                                let nested_items = items.clone();
+                                items = None;
+                                SimplifiedJsonSchema {
+                                    bson_type: Some(t),
+                                    items: nested_items,
+                                    ..Default::default()
+                                }
+                            }
+                            // If one of the BSON types is Object, move the
+                            // properties, required, and additional_properties
+                            // information into the schema that will be nested
+                            // in the top-level any_of and set the top-level
+                            // fields to None.
+                            BsonTypeName::Object => {
+                                let nested_properties = properties.clone();
+                                let nested_required = required.clone();
+                                let nested_additional_properties = additional_properties;
+                                properties = None;
+                                required = None;
+                                additional_properties = false;
+                                SimplifiedJsonSchema {
+                                    bson_type: Some(t),
+                                    properties: nested_properties,
+                                    required: nested_required,
+                                    additional_properties: nested_additional_properties,
+                                    ..Default::default()
+                                }
+                            }
+                            // If we encounter any other BSON type, simply create
+                            // a SimplifiedJsonSchema with the relevant bson_type.
+                            t => SimplifiedJsonSchema {
+                                bson_type: Some(t),
+                                ..Default::default()
+                            },
+                        }
+                    })
+                    .collect::<Vec<SimplifiedJsonSchema>>();
+
+                // Extend the top-level any_of with the new, additional
+                // schemas created here from the Multiple BSON types.
+                any_of = match any_of {
+                    None => Some(additional_any_of_schemas),
+                    Some(mut v) => {
+                        v.append(&mut additional_any_of_schemas);
+                        Some(v)
+                    }
+                };
+
+                // Set the top-level bson_type to None when there are multiple
+                // BSON types.
+                None
+            }
+        };
+
+        // Flatten nested any_ofs
```

Comment:
Deeply nested any_ofs are handled by the recursive calls of this method (via calls to `into()` earlier in the method).